pkg/hubble/exporter/aggregator.go

```diff
@@ -0,0 +1,106 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package exporter
+
+import (
+	"log/slog"
+
+	"google.golang.org/protobuf/proto"
+
+	flowpb "github.com/cilium/cilium/api/v1/flow"
+	v1 "github.com/cilium/cilium/pkg/hubble/api/v1"
+	fa "github.com/cilium/cilium/pkg/hubble/parser/fieldaggregate"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+// AggregateKey represents a combination of field values used for aggregation
+type AggregateKey string
+
+// AggregateValue holds the counters for ingress, egress, and unknown direction flows
+type AggregateValue struct {
+	IngressFlowCount          int
+	EgressFlowCount           int
+	UnknownDirectionFlowCount int
+	ProcessedFlow             *flowpb.Flow
+}
+
+type Aggregator struct {
+	m              map[AggregateKey]*AggregateValue
+	mu             lock.RWMutex
+	fieldAggregate fa.FieldAggregate
+	logger         *slog.Logger
+}
+
+func NewAggregator(logger *slog.Logger) *Aggregator {
+	a := &Aggregator{
+		m:      make(map[AggregateKey]*AggregateValue),
+		logger: logger.With(logfields.LogSubsys, "hubble-aggregator"),
+	}
+	return a
+}
+
+func NewAggregatorWithFields(fieldAggregate fa.FieldAggregate, logger *slog.Logger) *Aggregator {
+	a := &Aggregator{
+		m:              make(map[AggregateKey]*AggregateValue),
+		fieldAggregate: fieldAggregate,
+		logger:         logger.With(logfields.LogSubsys, "hubble-aggregator"),
+	}
+	return a
+}
+
+// Add processes an event directly into the aggregation map
+func (a *Aggregator) Add(ev *v1.Event) {
+	f := ev.GetFlow()
+	if f == nil {
+		return
+	}
+
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	// Create processed flow with only selected fields
+	processedFlow := &flowpb.Flow{}
+	a.fieldAggregate.Copy(processedFlow.ProtoReflect(), f.ProtoReflect())
+
+	// Generate key using the already processed flow
+	k := a.generateAggregationKey(processedFlow)
+
+	v, ok := a.m[k]
```

Comment:
```suggestion
	// Create processed flow with only selected fields
	processedFlow := &flowpb.Flow{}
	a.fieldAggregate.Copy(processedFlow.ProtoReflect(), f.ProtoReflect())

	// Generate key using the already processed flow
	k := a.generateAggregationKey(processedFlow)

	a.mu.Lock()
	defer a.mu.Unlock()

	v, ok := a.m[k]
```

suggestion/nit: I would move the locking lower to bypass steps that don't depend on protected fields and reduce the amount of time we're locked.